src/DependencyInjection/DI/src/ServiceLookup/Expressions/ExpressionResolverBuilder.cs

```diff
@@ -41,102 +40,84 @@ internal class ExpressionResolverBuilder : CallSiteVisitor<CallSiteExpressionBui
 
         private readonly ServiceProviderEngineScope _rootScope;
 
-        public ExpressionResolverBuilder(CallSiteRuntimeResolver runtimeResolver, IServiceScopeFactory serviceScopeFactory, ServiceProviderEngineScope rootScope):
-            base()
+        private readonly ConcurrentDictionary<ServiceCacheKey, Func<ServiceProviderEngineScope, object>> _scopeResolverCache;
+
+        public ExpressionResolverBuilder(CallSiteRuntimeResolver runtimeResolver, IServiceScopeFactory serviceScopeFactory, ServiceProviderEngineScope rootScope)
         {
             if (runtimeResolver == null)
             {
                 throw new ArgumentNullException(nameof(runtimeResolver));
             }
+
+            _scopeResolverCache = new ConcurrentDictionary<ServiceCacheKey, Func<ServiceProviderEngineScope, object>>();
             _runtimeResolver = runtimeResolver;
             _serviceScopeFactory = serviceScopeFactory;
             _rootScope = rootScope;
         }
 
         public Func<ServiceProviderEngineScope, object> Build(ServiceCallSite callSite)
         {
-            if (callSite.Cache.Location == CallSiteResultCacheLocation.Root)
+            // Only scope methods are cached
+            if (callSite.Cache.Location == CallSiteResultCacheLocation.Scope)
             {
-                // If root call site is singleton we can return Func calling
-                // _runtimeResolver.Resolve directly and avoid Expression generation
-                if (TryResolveSingletonValue(callSite, out var value))
-                {
-                    return scope => value;
-                }
-
-                return scope => _runtimeResolver.Resolve(callSite, scope);
+#if NETCOREAPP
+                return _scopeResolverCache.GetOrAdd(callSite.Cache.Key, (key, cs) => BuildNoCache(cs), callSite);
```

Comment:
Is it really worth it to conditionally use this overload? `this` is captured by closure either way. You could cache the closure though.

Reply:
You are right. https://github.com/dotnet/csharplang/issues/1277 would've helped.